DRILL-7672: Make metadata type required when reading from / writing into Drill Metastore

arina-ielchiieva opened the pull request:
# [DRILL-7672](https://issues.apache.org/jira/browse/DRILL-7672): Make metadata type required when reading from / writing into Drill Metastore

## Description

1. Upgraded Iceberg version and removed unneeded code for In / NotIn Expressions.
2. Updated Metastore Read / Modify interfaces to support required metadata types:
 a. introduced abstract Read / Modify classes with boilerplate code;
 b. added delete operation with filter and metadata type;
 c. added metadata type validator which checks supported metadata types for each component;
 d. made purge operation terminal;
 e. made necessary changes in REAME.md files.
3. Added / updated unit tests.

## Documentation
NA

## Testing
Added unit tests.

paul-rogers commented on metastore/iceberg-metastore/src/main/java/org/apache/drill/metastore/iceberg/components/tables/IcebergTables.java:
```diff
@@ -42,16 +45,15 @@
+    TableInfo.STORAGE_PLUGIN,
+    TableInfo.WORKSPACE,
+    TableInfo.TABLE_NAME,
+    MetadataInfo.METADATA_KEY);
```
Thanks for cleaning up this code. If these are the four types, does it make sense to define them in an enum, with the string keys as a property of the enum? Maybe as part of `MetadataType`?

arina-ielchiieva replied:
Actually this a good idea to pull all Metastore column names into one place. I have created `MetastoreColumn` enum where all columns and their names are stored. Also made sure that in the code we refer only to `MetastoreColumn` and not to its string representation.

paul-rogers commented on metastore/iceberg-metastore/src/test/java/org/apache/drill/metastore/iceberg/components/tables/TestIcebergTablesMetastore.java:
```diff
@@ -297,6 +302,7 @@ public void testOverwrite() {
       .workspace(tableInfo.workspace())
       .tableName(tableInfo.name())
       .metadataKey("dir0")
+      .metadataType(MetadataType.TABLE.name())
```
Some of these use the `MetadataType.TABLE`, others `MetadataType.TABLE.name()`. Do the tests make sure both work, or should they be uniform?

arina-ielchiieva replied:
In `TableMetadataUnit` most of the data is stored as string to be suitable for storage (for example, in the file), since this is unit test I had to create `TableMetadataUnit` manually and thus indicate metadata type string value, in the code we usually do not do this.

arina-ielchiieva commented:
@paul-rogers addressed code review comments. Had to do some refactoring to ensure we are always refer to `MetadataColumn` instead of its string representation, thus now PR contains slightly more files.
@vvysotskyi you might want to take a look as well since I have touched some new areas.

arina-ielchiieva commented:
@vvysotskyi thanks, @paul-rogers is the PR ok to go?
